File: design-patterns/ts/proxy/simple/proxy.ts
// proxy.ts

import Item from "./item";
import ProxyInterface from "./proxy-interface";

class Proxy implements ProxyInterface {
    private item: ProxyInterface | null = null;

    public operation1(): void {
        if (this.item == null) {
            this.item = new Item();
        }
        this.item.operation1();
    }

    public operation2(): void {
        if (this.item == null) {
            this.item = new Item();
        }
        this.item.operation2();
    }
}

export default Proxy;

File: design-patterns/go/proxy/simple/proxy.go
// proxy.go

package main

type Proxy struct {
	actual ProxyI
}

func NewProxy() (proxy *Proxy) {
	proxy = &Proxy{}
	return
}

func (proxy *Proxy) Operation1() {
	if proxy.actual == nil {
		proxy.actual = NewActual()
	}
	proxy.actual.Operation1()
}

func (proxy *Proxy) Operation2() {
	if proxy.actual == nil {
		proxy.actual = NewActual()
	}
	proxy.actual.Operation2()
}
